### ISSUE TITLE
Can't get the deploy to work

### ISSUE THREAD

**Issue by xannasavin** [sugatpoudel/git-deploy#6]
Hey, 
I tried to set this up, but I can't get it to work - I guess I am doing something wrong. Would you mind helping me?

My workflow file looks like this:

```yaml
# Controls when the action will run. 
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  push:
    branches: [ main ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  git-repo-deploy:
    name: 🎉 Deployment
    runs-on: ubuntu-latest
    steps:
    - name: 🚚 Deploy to REMOTE REPO
      uses: tagus/git-deploy@v0.4.1
      with:
        changes: mu-plugins
        repository: git@github.com:[USER]/[REMOTE REPO].git
        ssh_key: ${{ secrets.REPO_DEPLOY_KEY_REMOTE_REPO }}
        name: [My Name]
        email: [MY Email]
        clean_repo: true
````

Of course the [square brackets] are filled with my Data.
I'm trying to copy the whole Repo of this workflow to the Remote Repo, into the folder mu-plugins that is in the root of that remote repo. (More to follow, this is only the starting point)

It looks like everything runs until this error:

```
Checking out files: 100% (12227/12227), done.
copying changes
rsync: change_dir "/github/workspace//mu-plugins" failed: No such file or directory (2)
sending incremental file list
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1189) [sender=3.1.3]
```

I guess I'm misunderstanding the "changes" parameter?
(I hope the 12k does not really mean the amount of files, because then something is really wrong...)

Thanks a lot for a quick response!

**Comment by tagus** [comment]
hi @xannasavin,

from your error message, it looks like `mu-plugins` doesn't exist in your workspace; this action clones the specified git repo to a `./repo` folder in your workspace. The specified changes folder (`mu-plugins` in your case) is then copied over to `./repo`.

If you're trying to copy the entire workspace, you can specify `./` as the changes param. However note that this will copy everything to the root of the remote repo. If `mu-plugins` is a folder that exists in the remote repo and you only want to copy your changes there, this action doesn't directly support this however if your workspace looks like the following:

```
workspace
|_ changes_dir
  |_ mu-plugins
|_ repo
  |_ mu-plugins
  |_ ...
```

and you were to specify `changes_dir` as the `changes` param, then changes in `./changes_dir/mu-plugins` would be synced to `./repo/mu-plugins`. 

let me know if that addresses your issue or you're still running into problems

**Comment by xannasavin** [comment]
Hi @tagus,
thanks a lot for your reply. 
So I restructured the initial Repo so it contains the `mu-plugins` folder, but I still get the error
`rsync: change_dir "/github/workspace//mu-plugins" failed: No such file or directory (2)`
looking strange to me: the `//` between workspace and the `create` folder (which I specified without any `/` as seen above, but with the plain folders name only as seen in the examples of your readme)

**Comment by tagus** [comment]
can add this as a step before `git-deploy`; I want to ensure that it's not an invalid path issue

```
    steps:
    - name: output workspace
       run: ls
    - name: git-deploy
      uses: tagus/git-deploy@v0.3.2
      with:
         ...
```

if you can use `tree` (`tree -L 2`), that'd be even better but wasn't sure what base image you were using. also can you paste the params you're using for `git-deploy` (redact any personal info)

**Comment by xannasavin** [comment]
thank you!
Not sure what you are talking about "tree" - I'm pretty alpha when it comes to Github actions, so please bear with me!
So I downgraded to 0.3.2 (from the latest 0.4.1) and added the step. Below the output of the differen steps:

```
# Setup Job
1s
Current runner version: '2.303.0'
Operating System
Runner Image
Runner Image Provisioner
GITHUB_TOKEN Permissions
Secret source: Actions
Prepare workflow directory
Prepare all required actions
Getting action download info
Download action repository 'tagus/git-deploy@v0.3.2' (SHA:c7de901f0c7d498fefa2dbddf99290a5b3015550)
Complete job name: 🎉 Deployment

# Build tagus/git-deploy@v0.3.2
Build container for action use: '/home/runner/work/_actions/tagus/git-deploy/v0.3.2/Dockerfile'.
(not adding the whole output)

# output workspace
Run ls
  ls
  shell: /usr/bin/bash -e {0}

# Deploy to [SERVER]
Run tagus/git-deploy@v0.3.2
/usr/bin/docker run --name [****] --label 6c0442 --workdir /github/workspace --rm -e "INPUT_CHANGES" -e "INPUT_REPOSITORY" -e "INPUT_SSH_KEY" -e "INPUT_NAME" -e "INPUT_EMAIL" -e "INPUT_CLEAN_REPO" -e "INPUT_SSH_KNOWN_HOSTS" -e "INPUT_BRANCH" -e "INPUT_MESSAGE" -e "HOME" -e "GITHUB_JOB" -e "GITHUB_REF" -e "GITHUB_SHA" -e "GITHUB_REPOSITORY" -e "GITHUB_REPOSITORY_OWNER" -e "GITHUB_REPOSITORY_OWNER_ID" -e "GITHUB_RUN_ID" -e "GITHUB_RUN_NUMBER" -e "GITHUB_RETENTION_DAYS" -e "GITHUB_RUN_ATTEMPT" -e "GITHUB_REPOSITORY_ID" -e "GITHUB_ACTOR_ID" -e "GITHUB_ACTOR" -e "GITHUB_TRIGGERING_ACTOR" -e "GITHUB_WORKFLOW" -e "GITHUB_HEAD_REF" -e "GITHUB_BASE_REF" -e "GITHUB_EVENT_NAME" -e "GITHUB_SERVER_URL" -e "GITHUB_API_URL" -e "GITHUB_GRAPHQL_URL" -e "GITHUB_REF_NAME" -e "GITHUB_REF_PROTECTED" -e "GITHUB_REF_TYPE" -e "GITHUB_WORKFLOW_REF" -e "GITHUB_WORKFLOW_SHA" -e "GITHUB_WORKSPACE" -e "GITHUB_ACTION" -e "GITHUB_EVENT_PATH" -e "GITHUB_ACTION_REPOSITORY" -e "GITHUB_ACTION_REF" -e "GITHUB_PATH" -e "GITHUB_ENV" -e "GITHUB_STEP_SUMMARY" -e "GITHUB_STATE" -e "GITHUB_OUTPUT" -e "RUNNER_OS" -e "RUNNER_ARCH" -e "RUNNER_NAME" -e "RUNNER_TOOL_CACHE" -e "RUNNER_TEMP" -e "RUNNER_WORKSPACE" -e "ACTIONS_RUNTIME_URL" -e "ACTIONS_RUNTIME_TOKEN" -e "ACTIONS_CACHE_URL" -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/[repo_name]/[repo_name]":"/github/workspace" 6c0442:[****]
/root/.ssh does not exist, creating it
adding github.com to known_hosts
# github.com:22 SSH-2.0-babeld-81c37e11
adding ssh key
Identity added: (stdin) ((stdin))
updating git config
Cloning into 'repo'...
Warning: Permanently added the RSA host key for IP address '[****]' to the list of known hosts.

                                                                                
Checking out files:  71% (9173/12835)

[....]

Checking out files: 100% (12835/12835), done.
copying changes
rsync: change_dir "/github/workspace//mu-plugins" failed: No such file or directory (2)
sending incremental file list
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1189) [sender=3.1.3]

sent 20 bytes  received 12 bytes  64.00 bytes/sec
total size is 0  speedup is 0.00
```
and the yaml file:

```yaml
# Controls when the action will run. 
on:
  # Triggers the workflow on push or pull request events but only for the master branch
  # push:
    # branches: [ main ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  git-repo-deploy:
    name: 🎉 Deployment
    runs-on: ubuntu-latest
    steps:
    - name: output workspace
      run: ls
    - name: 🚚 Deploy to [other Repo]
      uses: tagus/git-deploy@v0.3.2
      with:
        changes: mu-plugins
        repository: git@github.com:[user]/[remote_repo].git
        ssh_key: ${{ secrets.REPO_DEPLOY_KEY_[remote_repo] }}
        name: [name]
        email: [email]
        clean_repo: true
```

**Comment by tagus** [comment]
looks like your workspace is completely empty as the `ls` step didn't seem to output anything so the `mu-plugins` folder does not exist hence the rsync error. This might be a setup issue, is `mu-plugins` a folder you create with a previous job/step or does it solely exist in your remote repo

**Comment by xannasavin** [comment]
Hi @tagus ,
no, the workspace is not empty 🤔 
not sure how to add screenshots here, but the repo consists of 2 folders and one file:

- .github/workflows
- mu-plugins
- .gitignore


the workflow of this repo is exactly as shown above.
the target repo also has the folder mu-plugins existing (among all the other folders the wp-content folder contains).

**Comment by xannasavin** [comment]
any news on this @tagus ?

**Comment by tagus** [comment]
hi @xannasavin, 

unsure how to proceed here since the `ls` output still says the workspace is empty ; I think i might be misunderstanding the configuration of the repo here (or there is a bug with the action). 

is it possible to create a minimalist repo that reproduces the problem you're seeing, would greatly help if i could reproduce it on my end

**Comment by xannasavin** [comment]
Hi @tagus ,
what would you need? I can invite you as a coworker to the repo I'm trying to get this to run, if that helps?

**Comment by tagus** [comment]
sure, would be happy to take a look

**Comment by xannasavin** [comment]
I added you to the repo as collaborator - let me know if you need anything else

**Comment by tagus** [comment]
finally got a chance to look at the repo; so the main issue was that the **checkout step** was missing 

```yaml
steps:
- uses: actions/checkout@v3 <--- was missing this step
- name: 🚚 Deploy to [other Repo]
   uses: tagus/git-deploy@v0.3.2
   // rest of action
```

github actions requires this since each job starts off with just the specified base image so this is a pre-requisite anytime you want to work with files in the repo (which is most of the time). more info [here](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions)

*I made a couple of commits trying to debug the workflow* 😅 

Also althougth the action was *successful*, you'll see that nothing was deployed since no changes were detected in the remote repo based on what was copied in from `mu-plugins`. 

When the `clean_repo: true` was provided to `git-deploy`, it tried to remove everything except what was in `mu-plugins` from the remote repo (which was most of this repo). This was causing some unknown error probably related to the large output volume. I would look into either ensuring that the necessary files are in the `mu-plugins` folder or deleting the unnecessary files from the remote repo manually before starting to rely on this deploy action.

Let me know if you run into any issues with `git-deploy` but hopefully the action itself should be fine.

**Comment by xannasavin** [comment]
thanks a lot for looking into this! I was not aware the checkout was missing, sorry!

so I agree "clean repo off" makes sense, since of course this repo should only be "added" to the existing repo, not deleting everything around.

In the remote repo I have currently just a dummy txt file in the folder mu-plugins, so the folder can exist. If I remove that, the folder is also deleted and the action does not re-create it.
does this mean, I need to include the whole code in mu-plugins in the remote repo and it will then be updated in the future using this action and repo? 
I tried to do so (added the whole mu-plugins folder and files to remote repo, did some minor changes to this repo´) but still "no changes detected, skipping push" (though there is one new file and also an additional line of code in the index.php).

**Comment by xannasavin** [comment]
Hey @tagus sorry for bothering, but any idea what might be the issue?

**Comment by tagus** [comment]
hey hey, your config looks fine but it's really weird that git is not detecting any changes ; I can confirm that there are definititely changes being copied in 🤔. As a sanity check can you try testing this out with a new remote repo. Basically create a branch new repo and set that as the `repository` and update the ssh key accordingly. 

Basically want to verify that it's not something with the `git-deploy` workflow. My guess is that something is off with the remote repo but we're definitelly glossing over some detail here.

Another approach would be try making changes locally to the remote repo and see if you see can see the changes with `git status`.

**Comment by xannasavin** [comment]
Hey thanks a lot for your help!
So I created a testrepo (added you there too) and it pushed the changes to the new repo (in the repo root, not into the mu-plugins folder, but it's something). Also an update push was transfered.
The question is now: why does it place the files in the repo root and not in the folder (that was of course not existing on first push)? and might this be related to the issue it is not capable of pushing to the other repo? it is not possible to target the "root" of the pushing repo instead of the folder and igore files like .gitignore and .github etc?

**Comment by tagus** [comment]
If i understand your intent correctly, you want the remote repo to have a top level `mu-plugins` folder with everything in the source `mu-plugins` directory. If this is the case, then the action won't support this with the current configuration since it copies everything from the specified `changes` directory to the remote repo root ([src](https://github.com/tagus/git-deploy/blob/c2e2f640875b181e0b048f115c3bc7dd8181dc28/entrypoint.sh#L51)).

You can structure your src repo to have a nested `mu-plugins` folder e.g.

```
.
|_ src
  |_ mu-plugins
    |_ ...
    |_ ...
    |_ ...
```

In the above example, you can specify `src` for the `changes` param and your remote repo should have `mu-plugins` be a top level folder.

**Comment by xannasavin** [comment]
Hi @tagus , sorry for the long period of silence - time was flying by.
I restructured the src repo a bit, so it matches the mu-plugins folder inside the root folder (since that repo is coming from a WordPRess installation itself I had to trick it a bit).
It looks now, that the commits are going through as expected 🎉 
So the source was the structure.
Thank you so much for your help! <3 
One last question: it looks like this action does not delete a file from the remote - I assume that's what you have to use the "cleanrepo" variable for? I'm a bit hesitating to activate that - will it just affect the remote folder mu-plugins or the whole repo?

**Comment by tagus** [comment]
It will end up affecting the entire remote repo since rsync will end up deleting everything outside of the `mu-plugins` folder in the remote repo, and a commit will be made in this state so the deployed repo will have a bunch of stuff deleted (which is probably not what you want).

Ideally, the src repo should mirror the remote repo so that we can use the `clean` flag but I think you were running into issues with this method before.

**Comment by tagus** [comment]
closing this thread but if you encounter any further issues, feel free to open a new issue 🙏🏽
